File: jsonxlsx/render.py
"""Template-based XLSX rendering with config-driven sheet strategies."""

import re
from copy import copy
from datetime import datetime, time, date as _date
from typing import Any

_DATE = re.compile(r"^\d{4}-\d{2}-\d{2}$")
_TIME = re.compile(r"^\d{2}:\d{2}:\d{2}$")


def coerce(v: Any) -> Any:
    """Coerce string values to date/time if they match ISO patterns."""
    if isinstance(v, str):
        if _DATE.match(v):
            return datetime.fromisoformat(v)
        if _TIME.match(v):
            return time.fromisoformat(v)
    return v


def copy_style(target: Any, template: Any) -> None:
    """Copy openpyxl cell styling from template to target cell."""
    target.font = copy(template.font)
    target.fill = copy(template.fill)
    target.border = copy(template.border)
    target.alignment = copy(template.alignment)
    target.number_format = template.number_format
    target.protection = copy(template.protection)


def render_value(
    directive: dict[str, Any], record: dict[str, Any], row_num: int, edges: dict[str, list[dict]] | None = None
) -> Any:
    """
    Apply a column directive to derive a cell value.

    Directives:
    - literal: constant value
    - formula: cell formula with {r} → row_num substitution
    - src: source field from record, optionally joined/templated
    - edge_lookup: join with edges (requires separate pass with edges context)
    """
    if "literal" in directive:
        return directive["literal"]

    if "formula" in directive:
        return directive["formula"].replace("{r}", str(row_num))

    if "edge_lookup" in directive:
        return None

    if "src" in directive:
        v = record.get(directive["src"])
        if v is None:
            return None
        if isinstance(v, list):
            sep = directive.get("join", "\n")
            tmpl = directive.get("item_template", "{x}")
            return sep.join(tmpl.format(n=i + 1, x=str(x)) for i, x in enumerate(v))
        return coerce(v)

    return None


def render_sheet_data_replace(
    ws: Any,
    sheet_cfg: dict[str, Any],
    masters: dict[str, list[dict]],
    edges: dict[str, list[dict]],
) -> None:
    """
    data_replace strategy: clear template data and inject from masters.

    Steps:
    1. Capture style template row
    2. Clear existing data rows (keep structure and styles)
    3. Inject master records sequentially (or pinned to xlsx_row if present)
    4. Apply edge_lookup joins in separate pass
    5. Apply bumon matrix if configured
    """
    source = sheet_cfg["source"]
    records = masters.get(source, [])
    data_start = sheet_cfg["data_start_row"]
    style_row = sheet_cfg.get("style_template_row", data_start)
    columns = sheet_cfg["columns"]
    bumon_matrix = sheet_cfg.get("bumon_matrix")

    cols_in_use: list[int] = [d["col"] for d in columns]
    if bumon_matrix:
        cols_in_use.extend(range(bumon_matrix["start_col"], bumon_matrix["end_col"] + 1))

    style_template = {c: ws.cell(style_row, c) for c in cols_in_use}

    for r in range(data_start, ws.max_row + 1):
        for c in cols_in_use:
            ws.cell(r, c).value = None
        ws.row_dimensions[r].height = None

    sequential_r = data_start
    for record in records:
        target_r = record.get("xlsx_row")
        if target_r is None:
            r = sequential_r
            sequential_r += 1
        else:
            r = target_r

        for directive in columns:
            cell = ws.cell(r, directive["col"], value=render_value(directive, record, r, edges))
            if directive["col"] in style_template:
                copy_style(cell, style_template[directive["col"]])

        for directive in columns:
            if "edge_lookup" not in directive:
                continue
            cfg = directive["edge_lookup"]
            edge_list = edges.get(cfg["source"], [])
            match_field = list(cfg["match"].keys())[0]
            record_field = cfg["match"][match_field]
            record_val = record.get(record_field)
            for e in edge_list:
                if e.get(match_field) != record_val:
                    continue
                if all(e.get(k) == v for k, v in cfg.get("where", {}).items()):
                    cell = ws.cell(r, directive["col"], value=e.get(cfg["select"]))
                    if directive["col"] in style_template:
                        copy_style(cell, style_template[directive["col"]])
                    break

        if bumon_matrix:
            bumon = masters.get("bumon", [])
            edge_list = edges.get(bumon_matrix["edge_source"], [])
            id_field = bumon_matrix["edge_id_field"]
            mark_field = bumon_matrix["edge_mark_field"]
            record_id = record.get("id")
            edge_map = {
                e["bumon_id"]: e.get(mark_field)
                for e in edge_list
                if e.get(id_field) == record_id
            }
            for b in bumon:
                col = bumon_matrix["start_col"] + b["order"] - 1
                if col > bumon_matrix["end_col"]:
                    continue
                cell = ws.cell(r, col, value=edge_map.get(b["id"]))
                if col in style_template:
                    copy_style(cell, style_template[col])


def render_preserve(ws: Any, sheet_cfg: dict[str, Any], masters: dict[str, list[dict]], edges: dict[str, list[dict]]) -> None:
    """
    preserve strategy: keep template as-is, optionally apply cell_overrides.

    Use when template already contains formatted data or complex layouts.
    """
    for ov in sheet_cfg.get("cell_overrides", []):
        ws[ov["cell"]] = ov["value"]

File: src/render.js
/**
 * Template-based XLSX rendering: load template, copy styles, inject data.
 * Supports data_replace (clear + inject) and preserve (keep template as-is) strategies.
 * Supports edge lookups (FK-like joins) and bumon matrix (N×M assignment grids).
 */

import ExcelJS from 'exceljs';

/**
 * Render preserve strategy: keep template as-is, optionally apply cell overrides.
 * @param {Object} config - Sheet config {sheet, cell_overrides}
 * @param {Object} state - {masters, edges}
 * @param {ExcelJS.Workbook} workbook - Loaded template workbook
 */
export async function renderPreserve(config, state, workbook) {
  const ws = workbook.getWorksheet(config.sheet);
  if (!ws) return;

  if (config.cell_overrides) {
    for (const override of config.cell_overrides) {
      const cell = ws.getCell(override.cell);
      cell.value = override.value;
    }
  }
}

/**
 * Render data_replace strategy: clear data rows, inject from masters, copy styles.
 * @param {Object} config - Sheet config {sheet, source, data_start_row, style_template_row, columns, edge_lookup, bumon_matrix}
 * @param {Object} state - {masters, edges}
 * @param {ExcelJS.Workbook} workbook - Loaded template workbook
 */
export async function renderSheetDataReplace(config, state, workbook) {
  const ws = workbook.getWorksheet(config.sheet);
  if (!ws) return;

  // Find source masters by type (if schema-driven)
  const sources = state.masters.filter((m) => {
    if (!config.source) return true;
    return m.type === config.source || m._type === config.source;
  });

  const startRow = config.data_start_row || 2;
  const styleRow = config.style_template_row || 2;
  const styleRowObj = ws.getRow(styleRow);

  // Clear existing data rows (preserve header)
  const maxRow = ws.rowCount || 1000;
  for (let r = startRow; r <= maxRow; r++) {
    ws.getRow(r).values = [];
  }

  // Inject masters
  let dataRow = startRow;
  for (const master of sources) {
    const rowObj = ws.getRow(dataRow);

    // Copy style from template row
    if (styleRowObj) {
      for (let col = 1; col <= (config.columns?.length || 50); col++) {
        const styleCell = styleRowObj.getCell(col);
        const targetCell = rowObj.getCell(col);
        if (styleCell.fill) targetCell.fill = { ...styleCell.fill };
        if (styleCell.font) targetCell.font = { ...styleCell.font };
        if (styleCell.border) targetCell.border = { ...styleCell.border };
        if (styleCell.alignment) targetCell.alignment = { ...styleCell.alignment };
      }
    }

    // Populate columns
    if (config.columns) {
      for (const col of config.columns) {
        const cell = rowObj.getCell(col.col);
        if (col.src) {
          cell.value = master[col.src] ?? '';
        } else if ('literal' in col) {
          cell.value = col.literal;
        }
      }
    }

    // Apply edge lookups (FK joins)
    if (config.columns) {
      for (const col of config.columns) {
        if (col.edge_lookup) {
          const lookup = col.edge_lookup;
          const edgeMatches = state.edges.filter((e) => {
            if (lookup.source && e._type !== lookup.source && e.type !== lookup.source) {
              return false;
            }
            if (lookup.match) {
              const hasMatch = Object.entries(lookup.match).every(
                ([edgeField, masterField]) =>
                  e[edgeField] === master[masterField]
              );
              if (!hasMatch) return false;
            }
            if (lookup.where) {
              const whereOk = Object.entries(lookup.where).every(
                ([field, val]) => e[field] === val
              );
              if (!whereOk) return false;
            }
            return true;
          });

          if (edgeMatches.length > 0) {
            const match = edgeMatches[0];
            const cell = rowObj.getCell(col.col);
            cell.value = match[lookup.select] ?? '';
          }
        }
      }
    }

    // Bumon matrix: N×M assignment grid
    if (config.bumon_matrix && state.edges) {
      const matrix = config.bumon_matrix;
      const idVal = master[matrix.id_field];
      const edgeMatches = state.edges.filter(
        (e) =>
          e[matrix.edge_id_field] === idVal &&
          (!matrix.source || e.type === matrix.source || e._type === matrix.source)
      );

      for (let colIdx = 0; colIdx < matrix.column_count; colIdx++) {
        const bumonId = matrix.bumon_ids[colIdx];
        const mark = edgeMatches
          .filter((e) => e[matrix.bumon_id_field] === bumonId)
          .map((e) => e[matrix.mark_field])
          .join('');
        const cell = rowObj.getCell(matrix.column_start + colIdx);
        cell.value = mark || '';
      }
    }

    dataRow++;
  }
}

/**
 * Load JSONL sheet configs and render each.
 * @param {string} configContent - JSONL content with sheet directives
 * @param {Object} state - {schema, masters, edges}
 * @param {ExcelJS.Workbook} workbook - Loaded template workbook
 */
export async function renderFromConfig(configContent, state, workbook) {
  const lines = configContent
    .split('\n')
    .map((l) => l.trim())
    .filter((l) => l);

  for (const line of lines) {
    const config = JSON.parse(line);
    if (!config.sheet) continue;

    const strategy = config.strategy || 'preserve';
    if (strategy === 'preserve') {
      await renderPreserve(config, state, workbook);
    } else if (strategy === 'data_replace') {
      await renderSheetDataReplace(config, state, workbook);
    }
  }
}

/**
 * Main render: load template XLSX, apply sheet configs, write output.
 * @param {Object} options - {templatePath, configContent, state, outputPath}
 */
export async function render(options) {
  const workbook = new ExcelJS.Workbook();

  // Load or create template
  if (options.templatePath) {
    await workbook.xlsx.readFile(options.templatePath);
  } else {
    // Create minimal template if none provided
    workbook.addWorksheet('Sheet1');
  }

  // If no template but we have config, create sheets from config
  if (!options.templatePath && options.configContent) {
    const lines = options.configContent
      .split('\n')
      .map((l) => l.trim())
      .filter((l) => l);
    const sheetNames = new Set();
    for (const line of lines) {
      const config = JSON.parse(line);
      if (config.sheet && !sheetNames.has(config.sheet)) {
        sheetNames.add(config.sheet);
        if (workbook.getWorksheet(config.sheet)) continue;
        const ws = workbook.addWorksheet(config.sheet);
        // Add header row with column placeholders
        if (config.columns) {
          for (const col of config.columns) {
            const cell = ws.getCell(1, col.col);
            cell.value = col.src || `Column ${col.col}`;
          }
        }
      }
    }
    // Remove default Sheet1 if we added other sheets
    if (sheetNames.size > 0) {
      const sheet1 = workbook.getWorksheet('Sheet1');
      if (sheet1) workbook.removeWorksheet(sheet1.id);
    }
  }

  // Apply sheet configs
  if (options.configContent && options.state) {
    await renderFromConfig(options.configContent, options.state, workbook);
  }

  // Write output
  if (options.outputPath) {
    await workbook.xlsx.writeFile(options.outputPath);
  }

  return workbook;
}
